Prevent sources assignment to default stock

aleron75 opened the pull request:
- change a previous fixture value to avoid conflict with specific test case introduced by these changes
- add a plugin to prevent assignment of sources to Default Stock
- add custom InvalidSourceAssignmentException
- update AssignSourcesToStockTest web API test

aleron75 committed:
Change a fixture value

This change is due to the fact that a test uses default stock id to check wrong source assignment;
only the default source shoud be assigned to the default stock and we will introduce a specific test for that
so this had to be changed.

aleron75 committed:
Change a fixture value

This change is due to the fact that a test uses default stock id to check wrong source assignment;
only the default source shoud be assigned to the default stock and we will introduce a specific test for that
so this had to be changed.

aleron75 committed:
Prevent assigning sources to Default Stocks

- add plugin to prevent assignment of sources to Default Stock
- add custom InvalidSourceAssignmentException
- update AssignSourcesToStockTest web api test

aleron75 committed:
Prevent assigning sources to Default Stocks

- add plugin to prevent assignment of sources to Default Stock
- add custom InvalidSourceAssignmentException
- update AssignSourcesToStockTest web api test

aleron75 committed:
Fix coding style issue

aleron75 commented:
The mess detector complains about unused `$subject` parameter in the before plugin but we can't avoid that parameter so what is the official guideline to deal with this kind of test failures?

naydav commented on app/code/Magento/InventoryCatalog/Plugin/InventoryApi/AssignSourcesToStock/PreventAssignSourcesToDefaultStock.php:
```diff
@@ -0,0 +1,54 @@
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryCatalog\Plugin\InventoryApi\AssignSourcesToStock;
```
This way has introduced broken modularity

Module `Inventory` has base implementation of Inventory business logic
This logic does not depends on concrete Catalog implementation

Сoncept of Default Stock is related to integration `Inventory` into current `Catalog`
And we have special module for this integration `InventoryCatalog`
So need to move this plugin (and declaration) in  `InventoryCatalog`

Desire state:
`Inventory` doesn't know about `Catalog`
`Catalog` doesn't know about `Inventory`
And `InventoryCatalog` module is bridge between `Catalog` and `Inventory`

One more example:
Symfony core and Doctrine
Symfony doesn't know about Doctrine
Doctrine doesn't know about Symfony
And DcotrineBundle  is bridge between Symfony and Doctrine

naydav replied:
And generally, we must not introduce plugins for the same module (plugin and object are in the same module)

aleron75 replied:
Hello @naydav sorry but I'm not sure to get the point here: 

>So need to move this plugin (and declaration) in InventoryCatalog

The plugin is **already** defined under the `InventoryCatalog` module.

Thank you.

naydav replied:
Sorry, It was my mistake

naydav commented on app/code/Magento/InventoryApi/Test/Api/StockSourceLink/AssignSourcesToStockTest.php:
```diff
@@ -135,12 +135,28 @@ public function dataProviderWrongParameters(): array
                     'soap_message' => 'Could not assign Sources to Stock',
                 ],
             ],
+            'multiple_sources_assigned_to_default_stock' => [
```
Need to move this check to `InventoryCatalog`
Yes, it will some code duplication but we need to do this for better modularity

naydav commented on app/code/Magento/InventoryCatalog/Exception/InvalidSourceAssignmentException.php:
```diff
@@ -0,0 +1,14 @@
+
+use Magento\Framework\Exception\InputException;
+
+class InvalidSourceAssignmentException extends InputException
```
Looks like we don't need to introduce new exception type because we don't have processing of this exception in our code.

For calls in client code, this exception will be wrapper in \Exception in `\Magento\Inventory\Model\StockSourceLink\Command\AssignSourcesToStock::execute`

naydav commented on app/code/Magento/InventoryCatalog/Plugin/InventoryApi/AssignSourcesToStock/PreventAssignSourcesToDefaultStock.php:
```diff
@@ -0,0 +1,54 @@
+use Magento\InventoryCatalog\Api\DefaultStockProviderInterface;
+use Magento\InventoryCatalog\Exception\InvalidSourceAssignmentException;
+
+class PreventAssignSourcesToDefaultStock
```
Pls add Plugin prefix for class name

naydav commented on app/code/Magento/InventoryCatalog/Plugin/InventoryApi/AssignSourcesToStock/PreventAssignSourcesToDefaultStock.php:
```diff
@@ -0,0 +1,54 @@
+     */
+    public function beforeExecute(AssignSourcesToStockInterface $subject, array $sourceIds, int $stockId)
+    {
+        if ($this->defaultStockProvider->getId() !== $stockId) {
```
Looks like we can merge this two conditions

naydav replied:
>The mess detector complains about unused $subject parameter in the before plugin but we can't > avoid that parameter so what is the official guideline to deal with this kind of test failures?
Need to introduce special annotation `@SuppressWarnings(PHPMD.UnusedFormalParameter)`

Example
https://github.com/magento-engcom/msi/blob/develop/app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/UpdateSourceItemAtLegacyQtyCounterPlugin.php#L87

naydav commented:
@aleron75 
>The mess detector complains about unused $subject parameter in the before plugin but we can't > avoid that parameter so what is the official guideline to deal with this kind of test failures?

Need to introduce special annotation `@SuppressWarnings(PHPMD.UnusedFormalParameter)`

Example
https://github.com/magento-engcom/msi/blob/develop/app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/UpdateSourceItemAtLegacyQtyCounterPlugin.php#L87

aleron75 committed:
Update source to implement changes proposed in PR review.

- Move some tests to InventoryCatalog
- Remove InvalidSourceAssignmentException
- Add missing Plugin prefix
- Merge two if conditions into one
- Introduce annotation to suppress UnusedFormalParameter warning

naydav commented:
@aleron75 
Looks like all is good

Only two things:
1) My comment related to API-functional test docblock
2) Need to merge develop branch and resolve conflict in `app/code/Magento/InventoryCatalog/etc/di.xml`

Thanks

naydav commented on app/code/Magento/InventoryCatalog/Test/Api/StockSourceLink/AssignSourcesToStockTest.php:
```diff
@@ -0,0 +1,91 @@
+    /**
+     * @magentoApiDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/sources.php
+     * @magentoApiDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/stock.php
+     * @param string|array $sourceIds
```
Looks like in our case `$sourceIds` can be the only array and `$stockId` can be the only int
Pls fix docblock and add type hinting for parameters
Thanks

aleron75 committed:
Update source to implement changes proposed in PR review.

- Merge branch 'develop' and solve conflicts in di.xml
- Update docblock and type hints of AssignSourcesToStockTest::testAssignSourcesToStockWithWrongParameters()

aleron75 committed:
Merge branch 'develop' into msi-prevent-sources-assignment-to-default-stock

aleron75 committed:
Fix missing closing tag in di.xml

aleron75 committed:
Fix code style (line too long)
